File: backend/services/image_validation.py
from io import BytesIO

from PIL import Image, UnidentifiedImageError
from pillow_heif import register_heif_opener

register_heif_opener(thumbnails=False)

MAX_IMAGE_BYTES = 5 * 1024 * 1024
MAX_IMAGE_DIMENSION = 4096
MAX_IMAGE_PIXELS = 16_777_216
SUPPORTED_IMAGE_FORMATS = {"JPEG", "PNG", "WEBP", "HEIF"}
SUPPORTED_FORMATS_MESSAGE = "Supported formats: JPG, PNG, WebP, and HEIC."
DECODE_ERROR_MESSAGE = "Unable to decode image. Use a valid JPG, PNG, WebP, or HEIC file."


def _check_dimensions(width: int, height: int) -> None:
    if (
        width <= 0
        or height <= 0
        or width > MAX_IMAGE_DIMENSION
        or height > MAX_IMAGE_DIMENSION
        or width * height > MAX_IMAGE_PIXELS
    ):
        raise ValueError("Image dimensions must be at most 4096 × 4096 pixels.")


def validate_image_bytes(image_bytes: bytes) -> bytes:
    """Validate file contents and return bytes OpenCV can decode for inference."""
    try:
        with Image.open(BytesIO(image_bytes)) as image:
            if image.format not in SUPPORTED_IMAGE_FORMATS:
                raise ValueError(SUPPORTED_FORMATS_MESSAGE)
            _check_dimensions(*image.size)
            image.verify()

        with Image.open(BytesIO(image_bytes)) as image:
            if image.format == "HEIF":
                output = BytesIO()
                image.convert("RGB").save(output, format="PNG")
                return output.getvalue()
        return image_bytes
    except (UnidentifiedImageError, OSError, Image.DecompressionBombError) as exc:
        raise ValueError(DECODE_ERROR_MESSAGE) from exc

File: deep_fe/frontend/src/utils/imageValidation.js
export const MAX_IMAGE_BYTES = 5 * 1024 * 1024;
export const MAX_IMAGE_DIMENSION = 4096;

const SUPPORTED_TYPES = new Set(["image/jpeg", "image/png", "image/webp", "image/heic", "image/heif"]);

export const isHeicFile = (file) =>
  ["image/heic", "image/heif"].includes(file.type) ||
  (!file.type && /\.(heic|heif)$/i.test(file.name));

export async function validateImageFile(file) {
  if (!SUPPORTED_TYPES.has(file.type) && !isHeicFile(file)) {
    throw new Error("Supported formats: JPG, PNG, WebP, and HEIC.");
  }
  if (!file.size) {
    throw new Error("The image file is empty.");
  }
  if (file.size > MAX_IMAGE_BYTES) {
    throw new Error("Image must be 5 MiB or smaller.");
  }

  // Browser image decoders do not consistently support HEIC; the API validates
  // its contents and dimensions before converting it for inference.
  if (isHeicFile(file)) return;

  let bitmap;
  try {
    bitmap = await createImageBitmap(file);
  } catch {
    throw new Error("Unable to decode image. Use a valid JPG, PNG, WebP, or HEIC file.");
  }

  try {
    if (
      !bitmap.width ||
      !bitmap.height ||
      bitmap.width > MAX_IMAGE_DIMENSION ||
      bitmap.height > MAX_IMAGE_DIMENSION
    ) {
      throw new Error("Image dimensions must be at most 4096 × 4096 pixels.");
    }
  } finally {
    bitmap.close();
  }
}
